# **Lab 01**

**Tên các thành viên**

<center>

| Họ tên | MSSV |
| ---- | ---- |
| Trần Nguyễn Nhật Cường | 22127048  |
| Nguyễn Công Tuấn | 22127436 |
| Trần Đăng Tuấn | 22127438 |

</center>

## **Cài đặt và sử dụng các thư viện cần thiết**

In [355]:
!pip install spotipy pandas numpy seaborn matplotlib

In [356]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import time

import spotipy
from spotipy.oauth2 import SpotifyClientCredentials
import csv

## **Thu thập dữ liệu**

**Nền tảng thu thập** - Spotify

**Phương pháp thư thập** - Gọi APIs được Spotify hỗ trợ

**Cách lưu trữ dữ liệu sau khi thu thập** - Lưu thành file vietnamese_songs.csv (comma-separated value) 

In [357]:
sp = spotipy.Spotify(auth_manager=SpotifyClientCredentials(
    client_id="f1225ca7d4d94873b8f8774b127aee92",
    client_secret="87130417da134688844d563a2e9e8ac7"
))

In [358]:
print(help(spotipy.Spotify))

Help on class Spotify in module spotipy.client:

class Spotify(builtins.object)
 |  Spotify(auth=None, requests_session=True, client_credentials_manager=None, oauth_manager=None, auth_manager=None, proxies=None, requests_timeout=5, status_forcelist=None, retries=3, status_retries=3, backoff_factor=0.3, language=None)
 |
 |  Example usage::
 |
 |      import spotipy
 |
 |      urn = 'spotify:artist:3jOstUTkEu2JkjvRdBA5Gu'
 |      sp = spotipy.Spotify()
 |
 |      artist = sp.artist(urn)
 |      print(artist)
 |
 |      user = sp.user('plamere')
 |      print(user)
 |
 |  Methods defined here:
 |
 |  __del__(self)
 |      Make sure the connection (pool) gets closed
 |
 |  __init__(self, auth=None, requests_session=True, client_credentials_manager=None, oauth_manager=None, auth_manager=None, proxies=None, requests_timeout=5, status_forcelist=None, retries=3, status_retries=3, backoff_factor=0.3, language=None)
 |      Creates a Spotify API client.
 |
 |      :param auth: An access token (

In [359]:
# vietnamese_playlists_id = [
#         "6XFOsAdp88ptBCdqUMAfmP",  # Nhạc Việt Hay Nhất Trên Spotify - VPOP 2021 On Spotify - Vpop Top Hits - Top 100 Việt Nam 
#         "0aiBKNSqiPnhtcw1QlXK5s",  # Nhạc Việt Top Hits - VPOP 2025
#         "3e7JoMenHvTB4sVE8fnn9O",  # nhạc chill Việt
#         "7tPGaU74ZtxIO9hsXVc9zj",  # Chill Vietnamese playlist
#         "2KfRwgF2GzDXHAPW1cCT7X",  # VPOP 2021-2022-2023
#         "02zSPmcae3ct4moT8om7kK",  # VPOP 2016-2017-2018
#         "4pkof2WT0Spca1pX3j2dh4",  # Vietnamese (V-Pop) 2025
#         "4ybaEp8KXd6YpHFy1npofA",  # VPop 2010-2016
#         "2U30HdoDyEZKr06144F4Jw",  # Vpop 2016-2019
#         "0ZOm8M5G8L7GHkTiU7DcMC",  # Vpop 2015s
#         "5f8z9oeanvcYu9ByRtoi9H",  # Vpop - Old but Gold
#         "6luqC1Np20Z0Ps576JFNj8",  # Nhạc Việt 2000s-2010s
#         "26OywSeeLgG0oJZlpq01nz",  # Vpop Làn sóng xanh for 8x 9x
#         "6DciQq1Mx1RKhlxNuKTn8q",  # Top Vpop VietNam
#         "56p6Zs4yC2ZVsHB6Dd9msj",  # Vpop 2020
#         "7p4UoGDCIhfTKavwQKTE2H",  # Vpop 2000-2011
#         "0mrqaVOe6KYEEcaSYhVK4a",  # VPOP TOP HITS Nhạc Việt Hay & Mới Nhất 2023 - Top 100+ Bài Hát Yêu Thích Được Nghe Nhiều Nhất
#         "2QL9xisvT5VOda2nGoH6QY",  # VPOP 2016 2017 2018     
#         "44oq55UFBOGqzbwAUzrGvM",  # VPOP Tiktok
#         "5qZ3WQIZBbaogqsLh6i5lF",  # VPOP playlist   
#         "4LzZcuYBacrMSX81LLvs9A",   # Top 50 VPOP nhiều lượt nghe nhất trên Spotify
#         "2yWOVvbnR2yPnVHwUnJjKW",  # VPOP 2013
#         "14Qvvt5OQng5ScEov3FpGW",  # VPOP 2014
#         "0d2Vu9QxnkmbNsBiecjGA1",  # VPOP 2015
#         "3ECBachqjVyTjkygCvBMN7",  # VPOP 2016
#         "2QOxomAadLMZ1Pr498bCP1",  # VPOP 2017
#         "2bFktF75SaQMWfJOjotEMP",  # VPOP 2018
#         "7LmFDhL7FmkHt1pSgcWqNy",  # VPOP 2019
#         "0ywpVJwBWNsmrG4nfGXOhs",  # VPOP 2020
#         "1cvnE797cGbHDeoz4uCaPd",  # VPOP 2021
#         "74nhehkDb5m0iaWYMUzrDQ",  # VPOP 2022
        
# ]


# def get_vietnamese_songs_from_playlists():
#     """Extract all Vietnamese songs from popular Spotify playlists in Vietnam."""
#     vietnam_songs = []

#     for playlist_id in vietnamese_playlists_id:
#         offset = 0
#         limit = 100  
#         while True:
#             results = sp.playlist_items(playlist_id, market="VN", limit=limit, offset=offset)
            
#             if not results or "items" not in results or not results["items"]:
#                 break 

#             for item in results["items"]:
#                 track = item.get("track")
                
#                 # If track is None, continue to next item
#                 if not track:
#                     continue
                
#                 album = track.get("album", {})

#                 vietnam_songs.append({
#                     "name": track.get("name", "null"),
#                     "release_date": album.get("release_date", "null"),
#                     "album_type": album.get("album_type", "null"),
#                     "album_name": album.get("name", "null"),
#                     "artists": ", ".join([artist["name"] for artist in track.get("artists", [])]) if track.get("artists") else "null",
#                     "duration_ms": track.get("duration_ms", "null"),
#                     "popularity": track.get("popularity", "null"),
#                     "spotify_url": track.get("external_urls", {}).get("spotify", "null"),
#                 })
            
#             offset += limit

#     return vietnam_songs

# def save_songs_to_csv(songs, filename="vietnamese_songs.csv"):
#     """Save song details into a CSV file."""
#     with open(filename, mode="w", newline="", encoding="utf-8") as file:
#         writer = csv.DictWriter(file, fieldnames=[
#             "name", "release_date", "album_type", "album_name",
#             "artists", "duration_ms", "popularity", 
#             "spotify_url"
#         ])
#         writer.writeheader()
#         for song in songs:
#             writer.writerow(song)

# vietnam_songs = get_vietnamese_songs_from_playlists()
# save_songs_to_csv(vietnam_songs)

## **Đọc bộ dữ liệu**

In [360]:
tracks = pd.read_csv(filepath_or_buffer='./vietnamese_songs.csv', encoding='utf-8')
tracks.head()

name release_date album_type  \
0           Chúng Ta Của Hiện Tại   2020-12-20     single   
1                       Đi Về Nhà   2020-12-20     single   
2  Sinh Ra Đã Là Thứ Đối Lập Nhau   2018-03-01     single   
3             Bước Qua Mùa Cô Đơn   2020-12-10     single   
4                        Nàng Thơ   2018-03-01     single   

                       album_name           artists  duration_ms  popularity  \
0           Chúng Ta Của Hiện Tại     Sơn Tùng M-TP       301538         0.0   
1                       Đi Về Nhà     Đen, JustaTee       200000        58.0   
2  Sinh Ra Đã Là Thứ Đối Lập Nhau  Emcee L (Da LAB)       234168         0.0   
3             Bước Qua Mùa Cô Đơn               Vũ.       278823        43.0   
4                        Nàng Thơ        Hoàng Dũng       254476         0.0   

                                         spotify_url  
0  https://open.spotify.com/track/17iGUekw5nFt5mI...  
1  https://open.spotify.com/track/7LVrrX8pkzI9fMF...  
2  https://open.spotify.com/track/6IAqflHsPVm4EpY...  
3  https://open.spotify.com/track/7wkX67ruxr3x2GF...  
4  https://open.spotify.com/track/2LykruN1mQlg9nu...

## **Khám phá dữ liệu**

Bảng bên dưới mô tả chi tiết ý nghĩa của các thuộc tính trong bộ dữ liệu `vietnamese_songs.csv`

<center>

| **Tên thuộc tính** | **Ý nghĩa** |
| -----              | -----       |
| **name** | Tên bài hát |
| **release_date** | Ngày bài hát được ra mắt/phát hành |
| **album_type** | Loại album mà bài hát thuộc về (ví dụ: album, single, compilation) |
| **album_name** | Tên của album mà bài hát thuộc về |
| **artists** | Tên của các nghệ sĩ trình diễn bài hát. Nếu có nhiều nghệ sĩ, tên các nghệ sĩ sẽ được phân tách bằng dấu phẩy (",") |
| **duration_ms** | Độ dài của bài hát tính theo đơn vị milli giây |
| **popularity** | Độ phổ biến của bài hát trên nền tảng Spotify, được tính từ 0 đến 100 (100 là phổ biến nhất) |
| **sportify_url** | Liên kết URL dẫn đến bài hát trên nền tảng Spotify |


</center>

## **Làm sạch dữ liệu**

### **Tổng quan về dữ liệu**

In [361]:
tracks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4120 entries, 0 to 4119
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   name          4117 non-null   object 
 1   release_date  4119 non-null   object 
 2   album_type    4119 non-null   object 
 3   album_name    4116 non-null   object 
 4   artists       4116 non-null   object 
 5   duration_ms   4120 non-null   int64  
 6   popularity    4119 non-null   float64
 7   spotify_url   4120 non-null   object 
dtypes: float64(1), int64(1), object(6)
memory usage: 257.6+ KB


**Nhận xét**:
- Dữ liệu có 4120 dòng và 8 thuộc tính (cột)
- Có một số cột bị thiếu giá trị bao gồm **`name`**, **`release_date`**, **`album_type`**, **`album_name`**, **`artists`** và **`popularity`**
- **release_data** cần chuyển sang kiểu dữ liệu **`datatime`**

In [362]:
tracks.describe()

duration_ms   popularity
count  4.120000e+03  4119.000000
mean   2.442462e+05    23.437728
std    5.293807e+04    20.937207
min    0.000000e+00     0.000000
25%    2.111680e+05     0.000000
50%    2.440000e+05    25.000000
75%    2.760000e+05    43.000000
max    1.453201e+06    69.000000

### **Thay đổi kiểu dữ liệu**

In [363]:
# Xử lý dữ liệu thiếu ngày hoặc tháng trong cột release_date
def fix_release_date(date):
    date = str(date).strip()  # Loại bỏ khoảng trắng thừa
    if date == "0" or date.lower() == "nan":  # Nếu là giá trị không hợp lệ
        return None
    elif len(date) == 4:  # Nếu chỉ có năm (YYYY)
        return date + "-01-01"  # Gán ngày mặc định 01-01
    return date  # Giữ nguyên nếu đã có đủ thông tin ngày-tháng-năm

# Áp dụng hàm xử lý
tracks['release_date'] = tracks['release_date'].astype(str).apply(fix_release_date)

# Chuyển thành kiểu datetime
tracks['release_date'] = pd.to_datetime(tracks['release_date'], errors='coerce')

### **Xử lý các bài hát bị trùng lặp**

In [364]:
tracks[tracks['name'] == "Nàng Thơ"]

name release_date album_type album_name     artists  duration_ms  \
4     Nàng Thơ   2018-03-01     single   Nàng Thơ  Hoàng Dũng       254476   
254   Nàng Thơ   2018-03-01      album         25  Hoàng Dũng       254476   
1575  Nàng Thơ   2018-03-01     single   Nàng Thơ  Hoàng Dũng       254476   
1855  Nàng Thơ   2020-03-08     single   Nàng Thơ  Hoàng Dũng       254476   
3492  Nàng Thơ   2018-03-01     single   Nàng Thơ  Hoàng Dũng       254476   
3676  Nàng Thơ   2020-12-16      album         25  Hoàng Dũng       254476   
3895  Nàng Thơ   2020-03-08     single   Nàng Thơ  Hoàng Dũng       254476   
4068  Nàng Thơ   2020-12-16      album         25  Hoàng Dũng       254476   

      popularity                                        spotify_url  
4            0.0  https://open.spotify.com/track/2LykruN1mQlg9nu...  
254          0.0  https://open.spotify.com/track/6rVaOdWvFaNyVVX...  
1575         0.0  https://open.spotify.com/track/2LykruN1mQlg9nu...  
1855        40.0  https://open.spotify.com/track/1s7d2cdaZuktw5Y...  
3492         0.0  https://open.spotify.com/track/2LykruN1mQlg9nu...  
3676        49.0  https://open.spotify.com/track/1w3eUC89GPspKpi...  
3895        40.0  https://open.spotify.com/track/1s7d2cdaZuktw5Y...  
4068        49.0  https://open.spotify.com/track/1w3eUC89GPspKpi...

**Nhận xét**:

Lấy ví dụ bài hát **Nàng thơ** của nghệ sĩ **Hoàng Dũng** được phát hành ở 2 thể loại khác nhau là đĩa đơn (Nàng thơ) và album (25) nên chúng ta cần loại bỏ trùng lặp dựa vào `name`, `artists`, `album_type` và `album_name`


In [366]:
tracks.drop_duplicates(subset=["name", "artists", "album_type", "album_name"], keep="last", inplace=True)

In [ ]:
tracks[tracks['name'] == "Nàng Thơ"]

name release_date album_type album_name     artists  duration_ms  \
3895  Nàng Thơ   2020-03-08     single   Nàng Thơ  Hoàng Dũng       254476   
4068  Nàng Thơ   2020-12-16      album         25  Hoàng Dũng       254476   

      popularity                                        spotify_url  
3895        40.0  https://open.spotify.com/track/1s7d2cdaZuktw5Y...  
4068        49.0  https://open.spotify.com/track/1w3eUC89GPspKpi...

### **Các câu hỏi**